primitives/trie/src/cache.rs

```diff
@@ -0,0 +1,794 @@
+	ForStorageRoot {
+		shared_value_cache: RwLockReadGuard<'a, NoHashingLruCache<CachedValue<H>>>,
+		shared_value_cache_access: MutexGuard<'a, IntSet<u64>>,
+		local_value_cache: MutexGuard<'a, IntMap<u64, CachedValue<H>>>,
+		storage_root: H,
+	},
+}
+
+impl<H: AsRef<[u8]>> ValueCache<'_, H> {
+	/// Get the value for the given `key`.
+	fn get(&mut self, key: &[u8]) -> Option<&CachedValue<H>> {
+		match self {
+			Self::Fresh(map) => map.get(key),
+			Self::ForStorageRoot {
+				local_value_cache,
+				shared_value_cache,
+				shared_value_cache_access,
+				storage_root,
+			} => {
+				let key = value_cache_get_key(key, &storage_root);
+
+				// We first need to look up in the local cache and then the shared cache.
+				// It can happen that some value is cached in the shared cache, but the
+				// weak reference of the data can not be upgraded anymore. This for example
+				// happens when the node is dropped that contains the strong reference to the data.
+				//
+				// So, the logic of the trie would lookup the data and the node and store both
+				// in our local caches.
+				local_value_cache.get(&key).or_else(|| {
+					shared_value_cache.peek(&key).map(|v| {
+						shared_value_cache_access.insert(key);
+						v
+					})
+				})
+			},
+		}
+	}
+
+	/// Insert some new `value` under the given `key`.
+	fn insert(&mut self, key: &[u8], value: CachedValue<H>) {
+		match self {
+			Self::Fresh(map) => {
+				map.insert(key.into(), value);
+			},
+			Self::ForStorageRoot { local_value_cache, storage_root, .. } => {
+				let key = value_cache_get_key(key, &storage_root);
+				local_value_cache.insert(key, value);
+			},
+		}
+	}
+}
+
+/// The actual [`TrieCache`](trie_db::TrieCache) implementation.
+///
+/// If this instance was created for using it with a [`TrieDBMut`](trie_db::TrieDBMut), it needs to
+/// be merged back into the [`LocalTrieCache`] with [`Self::merge_into`] after all operations are
+/// done.
+pub struct TrieCache<'a, H: Hasher> {
+	shared_node_cache: RwLockReadGuard<'a, SharedNodeCache<H>>,
+	shared_node_cache_access: MutexGuard<'a, HashSet<H::Out>>,
+	local_cache: MutexGuard<'a, HashMap<H::Out, NodeOwned<H::Out>>>,
+	value_cache: ValueCache<'a, H::Out>,
+}
+
+impl<'a, H: Hasher> TrieCache<'a, H> {
+	/// Merge this cache into the given [`LocalTrieCache`].
+	///
+	/// This function is only required to be called when this instance was created through
+	/// [`LocalTrieCache::as_trie_db_mut_cache`], otherwise this method is a no-op. The given
+	/// `storage_root` is the new storage root that was obtained after finishing all operations
+	/// using the [`TrieDBMut`](trie_db::TrieDBMut).
+	pub fn merge_into(self, local: &LocalTrieCache<H>, storage_root: H::Out) {
+		let cache = if let ValueCache::Fresh(cache) = self.value_cache { cache } else { return };
+
+		let mut value_cache = local.shared.value_cache.write();
+		cache
+			.into_iter()
+			.map(|(k, v)| (value_cache_get_key(&k, &storage_root), v))
```

Comment:
It probably doesn't matter in practice so you should probably ignore this, but technically we could just do something like this to avoid rehashing the storage root in a loop for each key (pseudocode; would also require that the order of hashing in `value_cache_get_key` is reversed):

```
let mut hasher = Hasher::new();
hasher.update(storage_root.as_ref()):
cache.into_iter().map(|(k, v)| {
    let mut hasher = hasher.clone();
    hasher.update(&k);
    (hasher.finish(), v)
})
```

Reply:
That also sounds like a good idea.